src/include/migraphx/op/unpack_fp4.hpp

```diff
@@ -0,0 +1,103 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_OPERATORS_UNPACK_FP4_HPP
+#define MIGRAPHX_GUARD_OPERATORS_UNPACK_FP4_HPP
+
+#include <migraphx/check_shapes.hpp>
+#include <migraphx/op/normalize_attribute.hpp>
+#include <migraphx/shape.hpp>
+#include <migraphx/config.hpp>
+#include <migraphx/par_for.hpp>
+#include <migraphx/argument.hpp>
+#include <migraphx/fp4_casts.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace op {
+struct unpack_fp4
+{
+    int64_t axis = -1;
+
+    std::string name() const { return "unpack_fp4"; }
+
+    value attributes() const
+    {
+        value normalize   = value::object{};
+        normalize["axis"] = value::array{normalize_attribute::include_min};
+        return {{"normalize_axes", normalize}};
+    }
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.axis, "axis"));
+    }
+
+    migraphx::shape normalize_compute_shape(std::vector<migraphx::shape> inputs) const
+    {
+        check_shapes{inputs, *this}.same_dims().has(1);
+        const auto& in_shape = inputs.front();
+        if(in_shape.type() != migraphx::shape::fp4x2_type)
+        {
+            MIGRAPHX_THROW("UNPACK_FP4: Only fp4x2_type is supported for unpacking");
+        }
+        auto new_lens = in_shape.lens();
+        new_lens[axis] *= 2;
+        return {migraphx::shape::float_type, new_lens};
+    }
+
+    argument compute(const shape& output_shape, const std::vector<argument>& args) const
+    {
+        const auto& input = args.front();
+        auto in_shape     = input.get_shape();
+
+        argument uint8_input        = input.reshape({migraphx::shape::uint8_type, in_shape.lens()});
+        migraphx::shape float_shape = shape{migraphx::shape::float_type, output_shape.lens()};
+        argument float_arg{float_shape};
+
+        float_arg.visit([&](auto out) {
+            uint8_input.visit([&](auto inp) {
+                par_for(in_shape.elements(), [&](auto i) {
+                    auto data_idx = in_shape.multi(i);
+                    data_idx[axis] *= 2;
+                    // unpacking 2 unsigned parts
+                    uint8_t val     = inp[i];
+                    uint8_t fp4_val = val & 0xFu;
```

Comment:
You already do this in `fp4_to_float` so this isnt necessary, unless you want to make this a precondition for `fp4_to_float`, which you should add an `assert` for that case then.